### PR TITLE
Narrower update to StreamingClientResult/SSE

### PR DESCRIPTION
**Pull request by trrwilson**

### CODE REVIEW

File: .dotnet/src/Custom/Chat/ChatClient.cs [hunk 1]
```diff
@@ -200,19 +200,27 @@ public virtual Task<ClientResult<ChatCompletion>> CompleteChatAsync(string messa
...
-        return StreamingClientResult<StreamingChatUpdate>.CreateFromResponse(
-            genericResult,
-            (responseForEnumeration) => SseAsyncEnumerator<StreamingChatUpdate>.EnumerateFromSseStream(
-                responseForEnumeration.GetRawResponse().ContentStream,
-                e => StreamingChatUpdate.DeserializeStreamingChatUpdates(e)));
+        PipelineResponse response = null;
+        try
+        {
+            response = requestMessage.ExtractResponse();
+            ClientResult genericResult = ClientResult.FromResponse(response);
```
Comment (KrzysztofCwalina) [inline .dotnet/src/Custom/Chat/ChatClient.cs:212]: Do you need this genericResult? It does not seem to be used anywhere.

File: .dotnet/src/Custom/Chat/StreamingChatUpdate.Serialization.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                            continue;
+                        }
+                        if (choiceProperty.NameEquals("finish_reason"u8))
+                        {
+                            if (choiceProperty.Value.ValueKind == JsonValueKind.Null)
+                            {
+                                finishReason = null;
+                                continue;
+                            }
+                            finishReason = choiceProperty.Value.GetString() switch
```
Comment (KrzysztofCwalina) [inline .dotnet/src/Custom/Chat/StreamingChatUpdate.Serialization.cs:80]: this is lower pri, but would it be better to stay in UTF8 span world and use a sequence of if statements (like for top level properties)

File: .dotnet/src/Utility/StreamingClientResult.cs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+namespace OpenAI;
+
+/// <summary>
+/// Represents an operation response with streaming content that can be deserialized and enumerated while the response
+/// is still being received.
+/// </summary>
+/// <typeparam name="T"> The data type representative of distinct, streamable items. </typeparam>
+public class StreamingClientResult<T>
+    : IDisposable
```
Comment (annelo-msft) [inline .dotnet/src/Utility/StreamingClientResult.cs:18]: Should not implement IDisposable - see discussion in https://github.com/Azure/azure-sdk-for-net/pull/43392 and related email thread.

File: .dotnet/src/Utility/StreamingClientResult.cs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    /// <param name="multiElementJsonDeserializerFunc">
+    /// The delegate that will be used to extract a collection of elements from each incoming JSON data payload.
+    /// </param>
+    /// <param name="cancellationToken">
+    /// The optional cancellation token used to control the enumeration.
+    /// </param>
+    /// <returns> A new instance of <see cref="StreamingClientResult{T}"/>. </returns>
+    public static StreamingClientResult<T> Create(
+        PipelineResponse response,
+        Func<JsonElement, IEnumerable<T>> multiElementJsonDeserializerFunc,
```
Comment (annelo-msft) [inline .dotnet/src/Utility/StreamingClientResult.cs:59]: We wouldn't take a Func in a ClientModel API - I found an alternate approach that works when you constrain T to IJsonModel<T> that is illustrated in https://github.com/joseharriaga/openai-in-typespec/pull/49

Comment (annelo-msft) [reply .dotnet/src/Utility/StreamingClientResult.cs:59]: Ah, just saw you added this below.

File: .dotnet/src/Utility/StreamingClientResult.cs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                    {
+                        yield return item;
+                    }
+                }
+            }
+        }
+        finally
+        {
+            // Always dispose the stream immediately once enumeration is complete for any reason
+            contentStream.Dispose();
```
Comment (annelo-msft) [inline .dotnet/src/Utility/StreamingClientResult.cs:125]: Disposal should be handled by the enumerator, not the enumerable.

File: .dotnet/src/Utility/StreamingClientResult.cs (hunk 4 above)
Comment (KrzysztofCwalina) [inline .dotnet/src/Utility/StreamingClientResult.cs:59]: Why do we need this parameter? Why the method below (with the constraint to IJsonModel) is not enoughl? In addition, we would need a Create method overload that returns StreamingClientResult<BinaryData> for the protocol method case, i.e. when the caller want to get access to raw bytes of each event data so they can deserialize additional properties themselves.

File: .dotnet/src/Utility/StreamingClientResult.cs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    /// <see cref="PipelineResponse.ContentStream"/>. This overload uses <see cref="ModelReaderWriter"/> via the
+    /// <see cref="IJsonModel{T}"/> interface and only supports single-item deserialization per server-sent event data
+    /// payload.
+    /// </summary>
+    /// <param name="response"> The base <see cref="PipelineResponse"/> for this result instance. </param>
+    /// <param name="cancellationToken">
+    /// The optional cancellation token used to control the enumeration.
+    /// </param>
+    /// <returns> A new instance of <see cref="StreamingClientResult{T}"/>. </returns>
+    public static StreamingClientResult<U> Create<U>(
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply .dotnet/src/Utility/StreamingClientResult.cs:78]: Strangely, can't apply `where` clauses without making the `method` specifically generic (`Create<T>`) and then making the method generic emits an analyzer warning that's effectively "hey, that's a different `T`, FYI."

File: .dotnet/src/Custom/Chat/ChatClient.cs (hunk 1 above)
Comment (KrzysztofCwalina) [reply .dotnet/src/Custom/Chat/ChatClient.cs:212]: but you are not returning this genericResult to the user, and so they cannot read the headers anyway.

File: .dotnet/src/Utility/StreamingClientResult.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    /// <see cref="PipelineResponse.ContentStream"/>. This overload uses <see cref="ModelReaderWriter"/> via the
+    /// <see cref="IJsonModel{T}"/> interface and only supports single-item deserialization per server-sent event data
+    /// payload.
+    /// </summary>
+    /// <param name="response"> The base <see cref="PipelineResponse"/> for this result instance. </param>
+    /// <param name="cancellationToken">
+    /// The optional cancellation token used to control the enumeration.
+    /// </param>
+    /// <returns> A new instance of <see cref="StreamingClientResult{T}"/>. </returns>
+    public static StreamingClientResult<U> Create<U>(PipelineResponse response, CancellationToken cancellationToken = default)
```
Comment (annelo-msft) [inline .dotnet/src/Utility/StreamingClientResult.cs:51]: Curious: is this used anywhere in the OAI client?

File: .dotnet/src/Utility/StreamingClientResult.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        where U : IJsonModel<U>
+    {
+        return new(response, GetServerSentEventDeserializationEnumerator<U>);
+    }
+
+    public static StreamingClientResult<TInstanceType> Create<TInstanceType, TJsonDataType>(PipelineResponse response, CancellationToken cancellationToken = default)
```
Comment (annelo-msft) [inline .dotnet/src/Utility/StreamingClientResult.cs:57]: Curious: is the `TJsonDataType` strictly required?  I notice that it is an internal collection type in your implementation.  Is it only to represent a collection?  Is the collection format part of the SSE spec?  If so, I am wondering if the factory method could just be `Create<T>` somehow and the collection could be left as an internal implementation detail.

**Pull request closed without merging**